packages/repository/src/relations/has-many/has-many-through.repository.ts

```diff
@@ -98,3 +105,129 @@ export interface HasManyThroughRepository<
     },
   ): Promise<void>;
 }
+
+export class DefaultHasManyThroughRepository<
+  TargetEntity extends Entity,
+  TargetID,
+  TargetRepository extends EntityCrudRepository<TargetEntity, TargetID>,
+  ThroughEntity extends Entity,
+  ThroughID,
+  ThroughRepository extends EntityCrudRepository<ThroughEntity, ThroughID>
+> implements HasManyThroughRepository<TargetEntity, TargetID, ThroughEntity> {
+  constructor(
+    public getTargetRepository: Getter<TargetRepository>,
+    public getThroughRepository: Getter<ThroughRepository>,
+    public getTargetConstraint: (
+      throughInstances: ThroughEntity[],
+    ) => DataObject<TargetEntity>,
+    public getThroughConstraint: (
+      targetInstance?: TargetEntity,
+    ) => DataObject<ThroughEntity>,
+  ) {}
+
+  async create(
+    targetModelData: DataObject<TargetEntity>,
+    options?: Options & {
+      throughData?: DataObject<ThroughEntity>;
+      throughOptions?: Options;
+    },
+  ): Promise<TargetEntity> {
+    const targetRepository = await this.getTargetRepository();
+    const throughRepository = await this.getThroughRepository();
+    const targetInstance = await targetRepository.create(
+      targetModelData,
+      options,
+    );
+    const throughConstraint = this.getThroughConstraint(targetInstance);
+    await throughRepository.create(
+      constrainDataObject(
+        options?.throughData ?? {},
+        throughConstraint as DataObject<ThroughEntity>,
+      ),
+      options?.throughOptions,
+    );
+    return targetInstance;
+  }
+
+  async find(
+    filter?: Filter<TargetEntity>,
+    options?: Options & {
+      throughOptions?: Options;
+    },
+  ): Promise<TargetEntity[]> {
+    const targetRepository = await this.getTargetRepository();
+    const throughRepository = await this.getThroughRepository();
+    const throughConstraint = this.getThroughConstraint();
+    const throughInstances = await throughRepository.find(
+      constrainFilter(undefined, throughConstraint),
+      options?.throughOptions,
+    );
+    const targetConstraint = this.getTargetConstraint(throughInstances);
+    return targetRepository.find(
+      constrainFilter(filter, targetConstraint),
+      options,
+    );
+  }
+
+  async delete(
+    where?: Where<TargetEntity>,
+    options?: Options & {
+      throughOptions?: Options;
+    },
+  ): Promise<Count> {
+    const targetRepository = await this.getTargetRepository();
+    const throughRepository = await this.getThroughRepository();
+    const throughConstraint = this.getThroughConstraint();
+    // TODO(derdeka): How to delete throughInstances?
```

Comment:
I think it should delete the through instance too. When the target is deleted, the FK in the through model will no longer reference a valid target model.

Reply:
On a relational database this can be easily done by a fk trigger `on delete cascade`. But what should happen on a non-relational database or if there is no trigger?

Reply:
IMO, this should delete through instances first and then delete the target instances second. That way:
1. We avoid violating referential integrity enforced by SQL databases by deleting target models that are still references by through models.
2. We avoid leaving orphaned "through" model instances in the databases. By an orphaned instance I mean an instance that's referencing a target model that no longer exists.

Please make sure to include acceptance-level test for this feature!

Here are two scenario to consider & cover by tests, I'll use "Source", "Target" and "Through" as model names instead of Customer/Seller/Order.

1. Given through links: `{sourceId: 1, targetId: 1}` and `{sourceId: 2, targetId: 2}`.
   When I delete all models belonging to source id 1,
   then the target model id 1 and the first link is deleted.

2. Given trough links `{sourceId: 1, targetId: 1}`, `{sourceId: 2, targetId: 1}` and `{sourceId: 2, targetId: 2}`. 
  When I delete all models belonging to source id 1, 
  then the target model id 1 and the first two links are deleted. Only `{sourceId: 2, targetId: 2}` link is preserved.